File: Sound_data/Demon_analysis/BroadBandAnalysis.py
import numpy as np
import scipy.signal as signal

def broadband_analysis(mixseg, noise, Fs, pars):
    BBsignal = []
    
    for j in range(len(pars['broadband']['FrequencyWindow'])):
        # Design the filter
        b, a = signal.butter(pars['broadband']['order'], 
                             np.array(pars['broadband']['FrequencyWindow'][j]) / Fs, 
                             btype='bandpass')

        # Apply the filter
        filtered_audio = signal.filtfilt(b, a, mixseg)
        filtered_noise = signal.filtfilt(b, a, noise)

        # Remove NaN and Inf values
        filtered_audio = filtered_audio[np.isfinite(filtered_audio)]
        filtered_noise = filtered_noise[np.isfinite(filtered_noise)]

        # Calculate broadband signal
        BBsignal.append(10 * np.log10(np.sum(np.abs(filtered_audio)**2) / 
                                      np.sum(np.abs(filtered_noise)**2)))

    return np.array(BBsignal)

File: data/Demon_matlab/BroadBandAnalysis.m
function[BBsignal] = BroadBandAnalysis(mixseg, noise, Fs, pars)

for j = 1:length(pars.broadband.FrequencyWindow(:,1)),
    
    % Filtering

    % Design the filter
    [b, a] = butter(pars.broadband.order, pars.broadband.FrequencyWindow(j,:)/Fs, 'bandpass');

    % Apply the filter to your audio data (assuming it's stored in a variable called 'audio_data')
    filtered_audio = filter(b, a, mixseg);
    filtered_noise = filter(b, a, noise);

    % RMS
    filtered_audio(isinf(filtered_audio)) = [];
    filtered_audio(isnan(filtered_audio)) = [];
    filtered_noise(isinf(filtered_audio)) = [];
    filtered_noise(isnan(filtered_audio)) = [];
    
    BBsignal(j) = 10*log10(sum(abs(filtered_audio).^2) / sum(abs(filtered_noise).^2));

end

return